### PR TITLE
Added GCDWebServerAsyncPushResponse response type

### PR DESCRIPTION
**Pull request by KJTsanaktsidis**
Continuing on from the discussion in #118 

In my project, I am implementing a proxy server which takes all requests, replays them to a remote server, and streams the response back to the original requester in my app. This PR adds a kind of buffered responder which you can push data into as you like, providing data to the web server when the appropriate dataReader callbacks are made. Hopefully this should address your concerns about choosing the right time to pass off the data to the  kernel.

I think this would be a good inclusion in the main project because other people doing similar things might find this mode of operation useful, but there is nothing special about this that requires it be in the GCDWebServerTree; I could easily just drop this code into my own project. 

What are your feelings about this kind of response type?

### CODE REVIEW

File: GCDWebServer/Responses/GCDWebServerAsyncPushResponse.m [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  }
+}
+
+- (void)pumpPendingRead {
+  if (_pendingReadBlock != nil) {
+    [self asyncReadDataWithCompletion:_pendingReadBlock];
+  }
+}
+
+- (void)sendWithData:(NSData *)data {
```
Comment (swisspol) [inline GCDWebServer/Responses/GCDWebServerAsyncPushResponse.m]: Don't use `@synchronized` it's very heavy handed, use a GCD serial queue instead.

File: GCDWebServer/Responses/GCDWebServerAsyncPushResponse.m [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  @synchronized(_buffer) {
+    [_buffer appendData:data];
+    [self pumpPendingRead];
+  }
+}
+
+- (void)completeWithErorr:(NSError *)error {
```
Comment (swisspol) [inline GCDWebServer/Responses/GCDWebServerAsyncPushResponse.m]: Typo

**Comment by swisspol** [comment]
I much prefer this approach indeed of putting the impedance logic entirely in the subclass. Does this build without access to the private headers?

**Comment by KJTsanaktsidis** [comment]
I have removed the use of @synchronize and fixed the typo. It does build without the private headers, it needs nothing more than the response protocol definition.

**Comment by KJTsanaktsidis** [comment]
@swisspol Just wondering if you're still thinking of merging this?

**Comment by swisspol** [comment]
I think that based on my comment at https://github.com/swisspol/GCDWebServer/pull/118#issuecomment-68586163, I will respectfully decline to merge this PR. The fact it can be just an extension to GCDWebServer in your code is also a viable workaround.

**Comment by KJTsanaktsidis** [comment]
OK, thank you for your help in suggesting this approach and reviewing my code. I will bring these extensions into our own project.

**Comment by KJTsanaktsidis** [comment]
In case anybody was interested in what I ended up putting into my project, it was this:
https://gist.github.com/KJTsanaktsidis/12ae946664fc3c1685d2

**Pull request closed without merging**